### ISSUE TITLE
Type promotion semantics not handled properly for `__torch_dispatch__`

### ISSUE THREAD

**Issue by makslevental** [pytorch/pytorch#74400]
### 🐛 Describe the bug

The TL;DR is that

```
Tensor[int] * ScalarTensor[float64] => Tensor[float64]
```
but
```
Tensor[int] * 2.0 => Tensor[float32]
```

The bug crops up in `__torch_dispatch__`, where when dispatching for `*` for something like `torch.randint(10, (2,2)) * 2.0` you get args with types `Tensor[int], ScalarTensor[float64]` and thus you end up having to return `Tensor[float64]`, whereas `aten` (i.e., vanilla PyTorch) will return `Tensor[float32]`. This is a problem for compiler backends that want to hook in through `__torch_dispatch__` (such as [torch-mlir](https://github.com/llvm/torch-mlir/pull/635)), since the compiler's type promotion semantics will disagree with PyTorch's. A straightforward repro of this using `functorch` is

```python
import torch
import torch.fx as fx
from functorch.compile import aot_module

def do_nothing_compiler():
    def f(fx_g: fx.GraphModule, inps):
        if hasattr(fx_g, "_tensor_constant0"):
            print(fx_g._tensor_constant0.dtype)
        return fx_g

    return f

class ElementwiseMulScalarModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return torch.mul(x, 8.0)

t = torch.ones((1, 3, 32, 32), dtype=torch.int32)
fn = aot_module(ElementwiseMulScalarModule(), do_nothing_compiler(), do_nothing_compiler())
y = fn(t)
print(y.dtype)
>>> torch.float64 # <---------- dtype that do_nothing_compiler sees for the scalar 8.0
>>> torch.float32 # <---------- dtype that PyTorch returns
```

I think this is related to [`wrapped_number`](https://github.com/pytorch/pytorch/blob/493bbdc4fea4f90f47ad94c693e3fafe0cede7b1/c10/core/TensorImpl.h#L1006-L1024) and [type promotion](https://github.com/pytorch/pytorch/blob/493bbdc4fea4f90f47ad94c693e3fafe0cede7b1/aten/src/ATen/native/TypeProperties.cpp#L102-L108) but I could be wrong.

cc @ezyang @gchanan @zou3519 @nairbv @mruberry @Chillee @albanD @samdow @silvasean @powderluv 

### Versions

```
OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.21.2
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.13.0-35-generic-x86_64-with-glibc2.31
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 510.39.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] functorch==0.2.0a0+6f3e137
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.12.0.dev20220316+cpu
[pip3] torchaudio==0.10.2
[pip3] torchvision==0.13.0.dev20220316+cpu
[pip3] torchviz==0.0.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] functorch                 0.2.0a0+1fa5434          pypi_0    pypi
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.12.0.dev20220316+cpu          pypi_0    pypi
[conda] torchaudio                0.10.2               py39_cu113    pytorch
[conda] torchvision               0.13.0.dev20220316+cpu          pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi
```

**Comment by Chillee** [comment]
This seems somewhat problematic to me, and also somewhat surprising 🤔 

I'm guessing there must be some wrapping going on when going back into python that messes something up?

**Comment by makslevental** [comment]
I put a breakpoint on [`ResultTypeState new_state = in_state;`](https://github.com/pytorch/pytorch/blob/493bbdc4fea4f90f47ad94c693e3fafe0cede7b1/aten/src/ATen/native/TypeProperties.cpp#L102) and ran

```python
import torch
torch.randint(10, (2,2)) * 2.0
```
and here's the stacktrace
`* thread #1, name = 'python', stop reason = breakpoint 1.1
    frame #0: libtorch_cpu.so`at::native::update_result_type_state(tensor=0xDEADBEEF, in_state=0xDEADBEEF) at TypeProperties.cpp:107:37
   104        current = typeMetaToScalarType(at::get_default_complex_dtype());
   105      }
   106      else if(isFloatingType(current)) {
-> 107        current = typeMetaToScalarType(at::get_default_dtype());
   108      }
   109    }
   110    if ( tensor.dim() > 0 ) {
(lldb) bt
* thread #1, name = 'python', stop reason = breakpoint 1.1
  * frame #0:  libtorch_cpu.so`at::native::update_result_type_state(tensor=0xDEADBEEF, in_state=0xDEADBEEF) at TypeProperties.cpp:107:37
    frame #1:  libtorch_cpu.so`at::TensorIteratorBase::compute_common_dtype(this=0xDEADBEEF) at TensorIterator.cpp:292:68
    frame #2:  libtorch_cpu.so`at::TensorIteratorBase::compute_types(this=0xDEADBEEF, config=0xDEADBEEF) at TensorIterator.cpp:427:41
    frame #3:  libtorch_cpu.so`at::TensorIteratorBase::build(this=0xDEADBEEF, config=0xDEADBEEF) at TensorIterator.cpp:1469:16
    frame #4:  libtorch_cpu.so`at::TensorIteratorBase::build_borrowing_binary_op(this=0xDEADBEEF, out=0xDEADBEEF, a=0xDEADBEEF, b=0xDEADBEEF) at TensorIterator.cpp:966:8
    frame #5:  libtorch_cpu.so`at::meta::structured_mul_Tensor::meta(this=0xDEADBEEF, self=0xDEADBEEF, other=0xDEADBEEF) at BinaryOps.cpp:62:28
    frame #6:  libtorch_cpu.so`at::(anonymous namespace)::wrapper_mul_Tensor(self=0xDEADBEEF, other=0xDEADBEEF) at RegisterCPU.cpp:6044:8
    frame #7:  libtorch_cpu.so`c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(const at::Tensor&, const at::Tensor&), at::(anonymous namespace)::wrapper_mul_Tensor>, at::Tensor, c10::guts::typelist::typelist<const at::Tensor&, const at::Tensor&> >, at::Tensor(const at::Tensor&, const at::Tensor&)>::call(c10::OperatorKernel *, c10::DispatchKeySet, const at::Tensor &, const at::Tensor &) [inlined] operator(args#1=0xDEADBEEF, args#0=0xDEADBEEF, this=0xDEADBEEF) at WrapFunctionIntoFunctor.h:13:72
    frame #8:  libtorch_cpu.so`c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(const at::Tensor&, const at::Tensor&), at::(anonymous namespace)::wrapper_mul_Tensor>, at::Tensor, c10::guts::typelist::typelist<const at::Tensor&, const at::Tensor&> >, at::Tensor(const at::Tensor&, const at::Tensor&)>::call(functor=0xDEADBEEF, (null)=(repr_ = 1), args#0=0xDEADBEEF, args#1=0xDEADBEEF) at make_boxed_from_unboxed_functor.h:424
    frame #9:  libtorch_cpu.so`at::Tensor c10::callUnboxedKernelFunction<at::Tensor, at::Tensor const&, at::Tensor const&>(unboxed_kernel_func=0xDEADBEEF, functor=0xDEADBEEF, dispatchKeySet=(repr_ = 1), (null)=0xDEADBEEF, (null)=0xDEADBEEF) at KernelFunction_impl.h:57:72
    frame #10: libtorch_cpu.so`at::Tensor c10::Dispatcher::redispatch<at::Tensor, at::Tensor const&, at::Tensor const&>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, at::Tensor const&)> const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&) const [inlined] at::Tensor c10::KernelFunction::call<at::Tensor, at::Tensor const&, at::Tensor const&>((null)=0xDEADBEEF, (null)=0xDEADBEEF, dispatchKeySet=(repr_ = 1), opHandle=0xDEADBEEF, this=0xDEADBEEF) const at KernelFunction_impl.h:67:140
    frame #11: libtorch_cpu.so`at::Tensor c10::Dispatcher::redispatch<at::Tensor, at::Tensor const&, at::Tensor const&>(this=0xDEADBEEF, op=0xDEADBEEF, currentDispatchKeySet=(repr_ = 1), (null)=0xDEADBEEF, (null)=0xDEADBEEF)> const&, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&) const at Dispatcher.h:557
    frame #12: libtorch_cpu.so`at::_ops::mul_Tensor::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&) [inlined] c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, at::Tensor const&)>::redispatch(args#1=0xDEADBEEF, args#0=0xDEADBEEF, currentDispatchKeySet=(repr_ = 1), this=<unavailable>) const at Dispatcher.h:419:126
    frame #13: libtorch_cpu.so`at::_ops::mul_Tensor::redispatch(dispatchKeySet=(repr_ = 1), self=0xDEADBEEF, other=0xDEADBEEF) at Operators_0.cpp:3850
    frame #14: libtorch_cpu.so`at::redispatch::mul(dispatchKeySet=(repr_ = 1), self=0xDEADBEEF, other=0xDEADBEEF) at RedispatchFunctions.h:4292:76
    frame #15: libtorch_cpu.so`operator(__closure=0xDEADBEEF) at VariableType_0.cpp:6885:78
    frame #16: libtorch_cpu.so`torch::autograd::VariableType::(anonymous namespace)::mul_Tensor(ks=(repr_ = 549755813889), self=0xDEADBEEF, other=0xDEADBEEF) at VariableType_0.cpp:6886:6
    frame #17: libtorch_cpu.so`c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(c10::DispatchKeySet, const at::Tensor&, const at::Tensor&), torch::autograd::VariableType::(anonymous namespace)::mul_Tensor>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, const at::Tensor&, const at::Tensor&> >, at::Tensor(c10::DispatchKeySet, const at::Tensor&, const at::Tensor&)>::call(c10::OperatorKernel *, c10::DispatchKeySet, const at::Tensor &, const at::Tensor &) [inlined] operator(args#2=0xDEADBEEF, args#1=0xDEADBEEF, args#0=(repr_ = 549755813889), this=0xDEADBEEF) at WrapFunctionIntoFunctor.h:13:72
    frame #18: libtorch_cpu.so`c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(c10::DispatchKeySet, const at::Tensor&, const at::Tensor&), torch::autograd::VariableType::(anonymous namespace)::mul_Tensor>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, const at::Tensor&, const at::Tensor&> >, at::Tensor(c10::DispatchKeySet, const at::Tensor&, const at::Tensor&)>::call(functor=0xDEADBEEF, dispatchKeySet=(repr_ = 549755813889), args#0=0xDEADBEEF, args#1=0xDEADBEEF) at make_boxed_from_unboxed_functor.h:441
    frame #19: libtorch_cpu.so`at::Tensor c10::callUnboxedKernelFunction<at::Tensor, at::Tensor const&, at::Tensor const&>(unboxed_kernel_func=0xDEADBEEF, functor=0xDEADBEEF, dispatchKeySet=(repr_ = 549755813889), (null)=0xDEADBEEF, (null)=0xDEADBEEF) at KernelFunction_impl.h:57:72
    frame #20: libtorch_cpu.so`at::_ops::mul_Tensor::call(at::Tensor const&, at::Tensor const&) [inlined] at::Tensor c10::KernelFunction::call<at::Tensor, at::Tensor const&, at::Tensor const&>((null)=0xDEADBEEF, (null)=0xDEADBEEF, dispatchKeySet=(repr_ = 549755813889), opHandle=0xDEADBEEF, this=0xDEADBEEF) const at KernelFunction_impl.h:67:140
    frame #21: libtorch_cpu.so`at::_ops::mul_Tensor::call(at::Tensor const&, at::Tensor const&) at Dispatcher.h:548
    frame #22: libtorch_cpu.so`at::_ops::mul_Tensor::call(at::Tensor const&, at::Tensor const&) [inlined] c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, at::Tensor const&)>::call(args#1=0xDEADBEEF, args#0=0xDEADBEEF, this=<unavailable>) const at Dispatcher.h:414
    frame #23: libtorch_cpu.so`at::_ops::mul_Tensor::call(self=0xDEADBEEF, other=0xDEADBEEF) at Operators_0.cpp:3844
    frame #24: libtorch_python.so`at::Tensor::mul(this=0xDEADBEEF, other=0xDEADBEEF) const at TensorBody.h:2804:72
    frame #25: libtorch_python.so`operator(__closure=0xDEADBEEF, self=0xDEADBEEF, other=0xDEADBEEF) at python_variable_methods.cpp:10975:26
    frame #26: libtorch_python.so`torch::autograd::THPVariable_mul(self_=0xDEADBEEF, args=0xDEADBEEF, kwargs=0xDEADBEEF) at python_variable_methods.cpp:10977:27
    frame #27: libtorch_python.so`torch::autograd::TypeError_to_NotImplemented_<torch::autograd::THPVariable_mul>(self=0xDEADBEEF, args=0xDEADBEEF, kwargs=0xDEADBEEF) at python_variable_methods.cpp:16280:23
    frame #28: python`method_vectorcall_VARARGS_KEYWORDS at descrobject.c:332

```

Not sure if that helps but at least it confirms (for me) that `wrapped_number`  is implicated.

**Comment by ezyang** [comment]
This is definitely a wrapped number problem. Eager mode will work OK because the tensor you get secretly has the wrapped number bit set but when you put it into AOTAutograd and get an FX graph that information gets lost. Not sure what to do about it.

**Comment by ezyang** [comment]
As a start, we could try to detect if a tensor is a wrapped number (I don't think we expose bindings for this at the moment) and then save it to a graph as a Python number instead of a tensor.

**Comment by makslevental** [comment]
> I don't think we expose bindings for this at the moment

As far as I can tell that is the case (it is not bound). If you point me to where you'd like for the binding to sit, I can submit a PR that wires it up.

> then save it to a graph as a Python number instead of a tensor.

I have no design opinions, but can just say that would be satisfactory for my use-case (which is not actually related to FX - that was just the easiest way to show a repro).

**Comment by ezyang** [comment]
Add a new underscored function to native_functions.yaml. Underscore because we need to decide what the permanent status of wrapped number is going to be.

**Comment by ezyang** [comment]
So I attempted to just have IValue to Python conversion turn a wrapped number tensor back into a number, as follows:

```
diff --git a/torch/csrc/jit/python/pybind_utils.h b/torch/csrc/jit/python/pybind_utils.h
index 942cf67618..a6d8bb1ea8 100644
--- a/torch/csrc/jit/python/pybind_utils.h
+++ b/torch/csrc/jit/python/pybind_utils.h
@@ -693,8 +693,24 @@ inline py::object toPyObject(IValue ivalue) {
     return py::none();
   } else if (ivalue.isTensor()) {
     auto tensor = std::move(ivalue).toTensor();
-    guardAgainstNamedTensor<at::Tensor>(tensor);
-    return py::cast(autograd::Variable(std::move(tensor)));
+    if (tensor.unsafeGetTensorImpl()->is_wrapped_number()) {
+      TORCH_INTERNAL_ASSERT(tensor.device().is_cpu());
+      auto scalar_type = tensor.scalar_type();
+      switch (scalar_type) {
+        case ScalarType::Bool: return py::cast(*tensor.data_ptr<bool>());
+        case ScalarType::Long: return py::cast(*tensor.data_ptr<int64_t>());
+        case ScalarType::Double: return py::cast(*tensor.data_ptr<double>());
+        case ScalarType::ComplexDouble: return py::cast(*tensor.data_ptr<c10::complex<double>>());
+        default:
+          TORCH_CHECK(false,
+              "Missing cases in 'toPyObject' wrapped number handling! Can't convert ",
+              scalar_type,
+              " to a Python object");
+      }
+    } else {
+      guardAgainstNamedTensor<at::Tensor>(tensor);
+      return py::cast(autograd::Variable(std::move(tensor)));
+    }
   } else if (ivalue.isStorage()) {
     return py::cast(ivalue.toStorage());
   } else if (ivalue.isDouble()) {
```

But this doesn't work, because when we go back to the torch.ops operator, you end up with this:

```
  File "/data/users/ezyang/pytorch-tmp/torch/_ops.py", line 48, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: aten::mul() Expected a value of type 'Tensor' for argument 'other' but instead found type 'float'.
Position: 1
Value: 0.017453292519943295
Declaration: aten::mul.Tensor(Tensor self, Tensor other) -> (Tensor)
Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)

----------------------------------------------------------------------
```

in other words, it's expecting a tensor... and we don't have one anymore, we have a float. Oops.

**Comment by ezyang** [comment]
I think my plan for now is to teach torch.ops parser how to wrap scalars. But I do agree that it would be better to have torch.ops create a full fledged ArgParser.

**Comment by albanD** [comment]
> I think my plan for now is to teach torch.ops parser how to wrap scalars.

This might be BC-breaking when calling the overload packet because some of the signature that used to raise errors (and thus be skipped) will now work?
That does sound like an ok thing though.

**Comment by ezyang** [comment]
Well... it depends on how we sorted the overloads for overload resolution. I'm YOLOing the test suite for now and we'll see if anything nasty shows up.

**Comment by ezyang** [comment]
This is fixed now!

**Comment by makslevental** [comment]
thanks @ezyang, sorry for letting this slip (since I offered to fix :/)
